File: packages/php/src/Seguranca/Ataques/AtaqueRotacional.php
<?php

namespace Application\Seguranca\Ataques;

use Application\Seguranca\AlvoCriptografico;
use Application\Seguranca\AtaqueInterface;
use Application\Seguranca\ResultadoAtaque;
use Application\Seguranca\SkipAtaqueException;

/**
 * Ataque rotacional/slide: se a cifra for simétrica a deslocamentos circulares
 * de key/iv, então `keystream(rot(key), rot(iv))` seria uma rotação de
 * `keystream(key, iv)` - uma estrutura explorável. Testa todas as rotações de
 * byte e também coincidência direta do keystream.
 */
class AtaqueRotacional implements AtaqueInterface
{
    public function __construct(private int $tamanho = 64)
    {
    }

    public function nome(): string
    {
        return 'Rotacional/slide (simetria por rotação)';
    }

    public function executar(AlvoCriptografico $alvo): ResultadoAtaque
    {
        $chave = $alvo->chaveDeTeste();
        $iv = random_bytes($alvo->tamanhoIv());
        $ks1 = $alvo->gerarKeystreamBruto($chave, $iv, 'enc', $this->tamanho);
        if ($ks1 === null) {
            throw new SkipAtaqueException('Alvo não expõe gerarKeystreamBruto.');
        }

        $coincidencias = [];
        for ($k = 1; $k < strlen($chave); $k++) {
            $ks2 = $alvo->gerarKeystreamBruto(
                $this->rotacionar($chave, $k),
                $this->rotacionar($iv, $k),
                'enc',
                $this->tamanho,
            );

            if ($ks2 === $ks1) {
                $coincidencias[] = "rotação $k: keystream idêntico";
                continue;
            }
            $alvoRot = $this->rotacionar(substr($ks1, 0, 32), $k);
            if (substr($ks2, 0, 32) === $alvoRot) {
                $coincidencias[] = "rotação $k: keystream rotacionado";
            }
        }

        $vulneravel = !empty($coincidencias);

        return new ResultadoAtaque(
            $this->nome(),
            vulneravel: $vulneravel,
            severidade: $vulneravel ? 'critica' : 'info',
            detalhes: $vulneravel
                ? 'Simetria rotacional encontrada: ' . implode('; ', array_slice($coincidencias, 0, 5))
                : 'Nenhuma das ' . (strlen($chave) - 1) . ' rotações de byte reproduziu o keystream',
            dados: ['coincidencias' => $coincidencias],
        );
    }

    private function rotacionar(string $buf, int $k): string
    {
        $n = strlen($buf);
        $out = '';
        for ($i = 0; $i < $n; $i++) {
            $out .= $buf[($i + $k) % $n];
        }
        return $out;
    }
}

File: packages/dart/lib/Seguranca/Ataques/AtaqueRotacional.dart
import 'dart:typed_data';

import 'package:criptografia/Seguranca/AlvoCriptografico.dart';
import 'package:criptografia/Seguranca/Ataque.dart';
import 'package:criptografia/Seguranca/ResultadoAtaque.dart';
import 'package:criptografia/Seguranca/Util.dart';

Uint8List _rotacionar(Uint8List buf, int k) {
  final n = buf.length;
  final out = Uint8List(n);
  for (var i = 0; i < n; i++) {
    out[i] = buf[(i + k) % n];
  }
  return out;
}

/// Ataque rotacional/slide: se a cifra for simétrica a deslocamentos circulares
/// de key/iv, então `keystream(rot(key), rot(iv))` seria uma rotação de
/// `keystream(key, iv)` - uma estrutura explorável. Testa todas as rotações de
/// byte e também coincidência direta do keystream.
class AtaqueRotacional extends Ataque {
  final int _tamanho;

  AtaqueRotacional({this._tamanho = 64});

  @override
  String nome() => 'Rotacional/slide (simetria por rotação)';

  @override
  ResultadoAtaque executar(AlvoCriptografico alvo) {
    final chave = toBytes(alvo.chaveDeTeste());
    final iv = randomBytes(alvo.tamanhoIv());
    final ks1 = alvo.gerarKeystreamBruto(chave, iv, 'enc', _tamanho);

    final coincidencias = <String>[];
    for (var k = 1; k < chave.length; k++) {
      final ks2 = alvo.gerarKeystreamBruto(
        _rotacionar(chave, k),
        _rotacionar(iv, k),
        'enc',
        _tamanho,
      );

      if (_iguais(ks2, ks1)) {
        coincidencias.add('rotação $k: keystream idêntico');
        continue;
      }
      // metade inicial de ks1 rotacionada deve bater com ks2 se houver simetria
      final alvoRot = _rotacionar(ks1.sublist(0, 32), k);
      if (_iguais(ks2.sublist(0, 32), alvoRot)) {
        coincidencias.add('rotação $k: keystream rotacionado');
      }
    }

    final vulneravel = coincidencias.isNotEmpty;

    return ResultadoAtaque(
      nome(),
      vulneravel,
      vulneravel ? Severidade.critica : Severidade.info,
      vulneravel
          ? 'Simetria rotacional encontrada: '
                '${coincidencias.take(5).join('; ')}'
          : 'Nenhuma das ${chave.length - 1} rotações de byte reproduziu o '
                'keystream',
      {'coincidencias': coincidencias},
    );
  }

  bool _iguais(Uint8List a, Uint8List b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }
}
